klippy/extras/display/uc1701.py

```diff
@@ -0,0 +1,180 @@
+# Support for UC1701 (128x64 graphics) LCD displays
+#
+# Copyright (C) 2018  Kevin O'Connor <kevin@koconnor.net>
+# Copyright (C) 2018  Eric Callahan  <arksine.code@gmail.com>
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
+import logging
+from font8x14 import VGA_FONT as CHAR_SET
+
+BACKGROUND_PRIORITY_CLOCK = 0x7fffffff00000000
+
+class UC1701:
+    char_right_arrow = '\x1a'
+    CURRENT_BUF, OLD_BUF = 0, 1
+    EMPTY_CHAR = (0, 32, 255)
+    def __init__(self, config):
+        printer = config.get_printer()
+        # pin config
+        ppins = printer.lookup_object('pins')
+        pins = [ppins.lookup_pin('digital_out', config.get(name + '_pin'))
+                for name in ['cs','a0']]
+        mcu = None
+        for pin_params in pins:
+            if mcu is not None and pin_params['chip'] != mcu:
+                raise ppins.error("uc1701 all pins must be on same mcu")
+            mcu = pin_params['chip']
+            if pin_params['invert']:
+                raise ppins.error("uc1701 can not invert pin")
+        self.pins = [pin_params['pin'] for pin_params in pins]
+        self.mcu = mcu
+        self.spi_oid = self.mcu.create_oid()
+        self.cs_oid = self.mcu.create_oid()
+        self.a0_oid = self.mcu.create_oid()
+        self.mcu.add_config_object(self)
+        self.glyph_buffer = []
+        self.spi_xfer_cmd = self.set_pin_cmd = None
+        self.vram = ([bytearray(128) for i in range(8)], 
+                    [bytearray('~'*128) for i in range(8)])
+    def build_config(self):
+        self.mcu.add_config_cmd(
+            "config_spi_without_cs oid=%d bus=%d mode=%d rate=%d shutdown_msg=" % (
+                self.spi_oid, 0, 0, 4000000))
+        self.mcu.add_config_cmd(
+            "config_digital_out oid=%d pin=%s value=%d default_value=%d max_duration=%d" % (
+                self.cs_oid, self.pins[0], 1, 1, 0))
+        self.mcu.add_config_cmd(
```

Comment:
Looking a little closer at this, the use of config_spi_without_cs could be problematic.  The issue is that the commands aren't guaranteed to be sent atomically - so, for example, an spi_send to an unrelated device might get sent between the uc1701 cs enable and the uc1701 cs disable.  Then, the command to the unrelated device would be read erroneously by the uc1701.

Does the display work if the regular config_spi command is used?  Also, the spec seems to say a 10mhz speed is valid - any reason to go with the slower 4mhz speed?

Reply:
Regular config_spi with spi_transfer should work.  

I couldn't find a recommended SPI speed in the datasheet, I must have missed it.   I can try 10mhz.  I'll make these changes and send them ASAP.

Reply:
FYI, on page 41 of the specification I have (under AC characteristics) I read the serial bus cycle time as min 100ns (with 2.5V-3.3V power).  So, I interpreted that as 10Mhz.

Thanks.
-Kevin